task/mgmt-gateway/src/mgs_handler.rs

```diff
@@ -152,4 +221,104 @@ impl SpHandler for MgsHandler {
             Err(ResponseError::Busy)
         }
     }
+
+    fn sys_reset_prepare(
+        &mut self,
+        _sender: SocketAddrV6,
+        _port: SpPort,
+    ) -> Result<(), ResponseError> {
+        ringbuf_entry!(Log::MgsMessage(MgsMessage::SysResetPrepare));
+        self.reset_requested = true;
+        Ok(())
+    }
+
+    fn sys_reset_trigger(
+        &mut self,
+        _sender: SocketAddrV6,
+        _port: SpPort,
+    ) -> Result<Infallible, ResponseError> {
+        if !self.reset_requested {
+            return Err(ResponseError::SysResetTriggerWithoutPrepare);
+        }
+
+        let jefe = task_jefe_api::Jefe::from(crate::JEFE.get_task_id());
+        jefe.request_reset();
+
+        // If `request_reset()` returns, something has gone very wrong.
+        panic!()
+    }
+}
+
+#[derive(Default)]
+struct UpdateBuffer {
+    total_length: usize,
+    bytes_written: usize,
+    current_block: ArrayVec<[u8; BLOCK_SIZE_BYTES]>,
+}
+
+impl UpdateBuffer {
+    fn ingest_chunk(
+        &mut self,
+        update_task: &Update,
```

Comment:
Dumb question: why does this need to be passed in as an argument, rather than being owned by the `UpdateBuffer`?  (That would make the name confusing, to be fair, but you could rename to something like `UpdateWorker`)

Reply:
It could be owned by the `UpdateBuffer`, but it's also used directly by `MgsHandler` (in `update_start()`), so I think the options are:

1. Both own a copy
2. `UpdateBuffer` owns it, and `MgsHandler` reaches it via `self.update_buffer.update...` when it needs it
3. `MgsHandler` owns it and passes it to `UpdateBuffer` in this method

I lightly prefer option 3, but if you strongly prefer one of the other two I wouldn't argue.

Reply:
Ah, I missed that it's also used by `MgsHandler`.  I don't feel strongly about it, so the current implementation is fine.